File: Data Structures & Algorithms/partition-equal-subset-sum/submission-0.java
class Solution {
    public boolean canPartition(int[] nums) {
        int sum = 0;
        for (int num : nums) {
            sum += num;
        }
        if (sum % 2 == 1) {
            return false;
        }
        int target = sum / 2;
        HashSet<Integer> dp = new HashSet<>();
        dp.add(0);
        for (int num : nums) {
            HashSet<Integer> temp = new HashSet<>();
            for (int i : dp) {
                if (i + num == target) {
                    return true;
                }
                temp.add(i + num);
            }
            for (int i : temp) {
                dp.add(i);
            }
        }
        return false;
    }
}

File: Data Structures & Algorithms/find-target-in-rotated-sorted-array/submission-0.py
class Solution:
    def search(self, nums: List[int], target: int) -> int:
        l = 0
        r = len(nums) - 1

        while l < r:
            m = l + (r - l) // 2
            left = nums[l]
            right = nums[r]
            mid = nums[m]

            if mid > right:
                l = m + 1
            else:
                r = m

        if target <= nums[-1]:
            r = len(nums) - 1
        else:
            r = l - 1
            l = 0

        print(f"l: {l}, r: {r}")
        while l <= r:
            m = l + (r - l) // 2
            if nums[m] == target:
                return m
            elif nums[m] < target:
                l = m + 1
            else:
                r = m - 1
        return -1